src/FileSystem-Disk/DiskStore.class.st

```diff
@@ -79,12 +85,46 @@ DiskStore class >> maxFileNameLength [
 	self subclassResponsibility 
 ]
 
+{ #category : #public }
+DiskStore class >> readOnlyVariant [
+
+	"return a class variant for read-only store. It should be properly defined in my subclasses"
+
+	^ self subclassResponsibility
+]
+
 { #category : #current }
 DiskStore class >> reset [
 	DefaultWorkingDirectory := nil.
 	CurrentFS := nil
 ]
 
+{ #category : #current }
+DiskStore class >> resetAsReadOnly [
+
+	"Reset the disk store as read-only. Change indentity of the old store to utilize it in the old users"
+
+	| old |
+	
+	old := CurrentFS.
+	CurrentFS := FileSystem store: self activeClass readOnlyVariant createDefault.
+	old becomeForward: CurrentFS.
```

Comment:
I don't think that this functionality really requires special #become: feature from the language.
We don't need to replace existing FileSystem instance if it exist. we can just ask to to become readonly:
```Smalltalk
self currentFileSystem beReadOnly
```
And #beReadOnly will do the work:
```Smalltalk
FileSystem>>beReadOnly
   store := store asReadOnlyStore
```

Reply:
> And I agree with @gcotelli . It could be ReadOnlyStore , a wrapper over existing store.
> For your argument it's still can ask a store for #readOnlyVariant. But it will be common implementation. And for very special store implementation it could override this method for specific read only option.

I tried to implement it. We have two options. One is to use MNU mechanism which I would like to avoid if it is not really necessary. The second option is to implement the methods explicitly with the resending to internal store. Which means to resend not only DiskStore methods (57) but FileSysstemStore methods (71). That would be horrible mess, definitelly worse than in case of separate read-only store classes. And you would mix two design patterns together. The optimal solution would be a Trait usage but we cannot use it